enhancements/ignition-spec-dual-support.md

```diff
@@ -0,0 +1,327 @@
+means that UPI clusters are at risk when upgrading to an MCO with dual support.
+Mitigation methods are discussed below.
+
+
+### RHCOS
+
+The RHCOS bootimage needs to be updated to ignition package v2.0+ . Required
+dependencies are discussed below.
+
+
+### Installer
+
+The installer needs to be updated to use ignition spec 3 alongside switching
+the pinned RHCOS bootimage. This means it would need to switch all typing as
+well as generated configs. 
+
+
+### MCO
+
+The MCO and its subcomponents are the most affected by this change. The
+aforementioned spec translator will be housed in the MCO. This means that the
+MCO would need to simultaneously vendor both ignition v0 and v2, and translate
+existing spec 2 configs to spec 3. The translation will happen as soon as the
+support for ignition spec 3 lands in MCO. This means for existing machines,
+the existing rendered machineconfig will be translated to ignition spec 3 in
+the MCC. The MCS will have additional functionality to translate the existing
+ignition configs to spec 3. The MCS will also be able to detect which ignition
+version to serve a newly booted image for initial provisioning of the system.
+Then it will update the new node to align with the rest of the cluster, using
+spec 3. New machineconfigs applied to the cluster that has an ignition snippet
+on spec 2 should no longer be allowed. 
+
+The MCO will also throw an alert upon:
+ - An attempted to update to the new ignition spec cannot be performed due to untranslatable configs, and the user must manually update/remove the untranslatable config before the update can continue
+ - The cluster sees a spec 2 machineconfig applied to it after already updating to spec 3, and rejects that machineconfig.
+
+The MCO should also add ability to reconcile broken spec 2 -> spec 3 updates,
+after manual intervention from the user.
+
+
+### User Stories
+
+** As the admin of an existing 4.x cluster on spec 2, I’d like to upgrade to the newest version and use ignition spec 3 **
+
+Acceptance criteria:
+ - The update completes without user intervention, if all machine configs existing on the cluster can be directly translated to spec 3
+ - The user receives an alert, if the update is unable to complete due to untranslatable configs
+ - The user is able to recover the cluster and finish the update if the untranslatable configs are manually translated or remove, or roll back to the old version
+ - The user should have received notification that the update will be changing spec version, as well as received necessary documentation on how to recover a failed update
+ - CI tests are put in place to make sure the existing versions can be updated to the new payload
+
+** As a user of openshift, I’d like to install from a spec 2 bootimage andimmediately update to a spec 3 payload **
+
+Acceptance criteria:
+ - Essentially the same as story 1
+
+** As a user of Openshift, I’d like to install a fresh ignition spec 3 cluster **
```

Comment:
the users just want to install a cluster, imo there is less value in customers asking for spec2 vs spec3

above we mention that the installer can switch to spec3 completely, do we have some idea how we can make sure UPI customers are using the correct boot-images when they get the spec3 ignition configs but maybe try to use a spec3 supporting bootimage

Reply:
I agree, the workflow should look the same for IPI

> do we have some idea how we can make sure UPI customers are using the correct boot-images when they get the spec3 ignition configs but maybe try to use a spec3 supporting bootimage

We'll definitely have to communicate in the docs which preview version has spec 3 support. The installer maybe should add capability to detect if the user is still on an old version and reject the install.

Reply:
> The installer maybe should add capability to detect if the user is still on an old version and reject the install.

at what stage do you think the installer will be able to do that?

in UPI, the installer only outputs ignition files and rest is user controlled process.

Reply:
Oh true, I guess in this case the best we can do is to alert the user that this build will output spec 3 ignition configs, which is only supported by RHCOS versions x and later?

This goes back to https://github.com/openshift/installer/issues/1399 which might be a better UPI workflow. Alternatively, for example if we decide to do this in 4.5, all 4.5 images + installer generated configs are spec 3 by default only. In that case a user would only run in to a version mistmatch if they try to install a 4.5 cluster with 4.4 bootimages, which they shouldn't do anyways.

Reply:
Open to suggestions of course

Reply:
> This goes back to openshift/installer#1399 which might be a better UPI workflow. Alternatively, for example if we decide to do this in 4.5, all 4.5 images + installer generated configs are spec 3 by default only. In that case a user would only run in to a version mistmatch if they try to install a 4.5 cluster with 4.4 bootimages, which they shouldn't do anyways.

> if they try to install a 4.5 cluster with 4.4 bootimages, which they shouldn't do anyways.

that means that we don't have N-1 compatibility, which is not we tell our users! the n-1 should be fine as long as you don't depend on a feature that is in N.

> This goes back to openshift/installer#1399 which might be a better UPI workflow.

I have specifically pushed back on that idea because, that allows us to demand this 4.4 works with 4.4 bootimage only, when that shouldn't be the case for most users. and also something we need to support in our clusters because 4.1 created clusters have new nodes join in at 4.1 and then pivot to say 4.4.

> Open to suggestions of course since UPI in general will be harder to coordinate

+1
as long as we explicitly mention what we are doing for this case, i'm fine.